{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 2: What do the SARS-Cov-2 genes do?\n",
    "\n",
    "In the first homework, we wrote an algorithm to retrieve ORFs from the SARS-CoV-2 genome. Some retrieved ORFs may correspond to the actual genes, some were there by chance, and some because the start codon also encodes for methionine, so the retrieved ORF was a (short) part of the more extended coding sequence. We used protein hydrophobicity to find potential transmembrane proteins. But what about other types of proteins? How can we tell which ORFs correspond to the actual genes?\n",
    "\n",
    "We will answer the question above in the bioinformatics way and lean to what we know about evolution. If two species are morphologically different, their genetic material must also differ substantially. Right? Not exactly. Evolution conserved many genes crucial for survival. The degree of conservation is surprising. For example, while we would expect to share many genes with the chimpanzee, our closest relative, we would expect almost no similarity in our genetic material with bananas. The numbers tell us otherwise. While sharing 96% of our genes with the chimpanzee, we also share 60% of our genes with bananas! These similar genes in humans and bananas encode proteins that carry similar functions.\n",
    "\n",
    "Two genes with similar sequences may perform similar functions. Genes like these are called homologous genes. Or, more precisely, in our case, these would be orthologous genes (see the figure below). A pair of orthologous genes are genes from two related species originating from a common ancestor (e.g., humans and chimpanzees). Due to the evolution, the two nucleotide or protein sequences might differ slightly but not overwhelmingly.\n",
    "\n",
    "<div>\n",
    "<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/d/d8/Ortholog_paralog_analog_%28homologs%29.svg/1200px-Ortholog_paralog_analog_%28homologs%29.svg.png\" width=\"500\">\n",
    "</div>\n",
    "\n",
    "*Author: Thomas Shafee, Licence: CC BY 4.0*\n",
    "\n",
    "The human-banana example demonstrates that these two organisms don't have to be that closely related! If we tried to determine the gene function in humans using bananas as a reference, we could infer the role of over 60% of the genes. However, using two closely related organisms would undoubtedly give us more robust results. We would be better off choosing the genome of a mouse or a chimpanzee as our reference.\n",
    "\n",
    "Scientists have studied hundreds of viruses and determined the functions of most of their genes. Using viruses similar in sequence to SARS-Cov-2, we can characterize the SARS-CoV-2 ORFs we found in our first homework and determine if they are the actual genes, and think about their function. Knowing what each gene does might help us find a way to fight the infection and potentially even find treatments for COVID-19. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In order to import from the python file without hassle, we add the current\n",
    "# directory to the python path\n",
    "import sys; sys.path.append(\".\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's let the nice folks at NCBI know who we are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"tm09219@student.uni-lj.si\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding SARS-CoV-2's closest relatives\n",
    "\n",
    "In the first part of this assignment, we will analyze several coronavirus genomes to determine which are most similar to SARS-CoV-2. We will then use the viruses with the most similar sequence as reference genomes to assign functions to SARS-CoV-2 ORFs.\n",
    "\n",
    "To find the closest reference genomes, we will align the SARS-CoV-2 sequence to each candidate coronavirus genome. The virus genomes with the highest alignment scores will be our most closely-related viruses.\n",
    "\n",
    "Notice that computing an alignment between two sequences of length $N$ and $M$ requires the computation of a dynamic programming table with $N \\cdot M$ entries. The size of this matrix is sufficiently small for short sequences. But even the short genomes, like viral genomes, are generally too long for this approach. Instead of computing similarities from the entire genomes, we will only focus on the spike protein sequence. In general, we would prefer to align the whole nucleotide sequences, but for this assignment, considering only a spike protein will be sufficient. The spike protein is also one of the essential parts of any coronavirus, as it is the one that grants the virus entry to host cells. \n",
    "\n",
    "Considering spike proteins alone reduces the sequence lengths from ~30k (entire virus) to around 1.3k (spike protein). Nevertheless, do your best to write fast, efficient Python code. On the laptop of your teaching assistants, each `global_alignment` call on 1.3k long protein sequences takes around 30 to 60 seconds. We have to calculate 20 comparisons, which takes roughly 10 to 20 minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "accession_codes = {\n",
    "    # 6 known human coronaviruses\n",
    "    \"Human-SARS\": \"NC_004718\",\n",
    "    \"Human-MERS\": \"NC_019843\",\n",
    "    \"Human-HCoV-OC43\": \"NC_006213\",\n",
    "    \"Human-HCoV-229E\": \"NC_002645\",\n",
    "    \"Human-HCoV-NL63\": \"NC_005831\",\n",
    "    \"Human-HCoV-HKU1\": \"NC_006577\",\n",
    "    \n",
    "    # Bat\n",
    "    \"Bat-CoV MOP1\": \"EU420138\",\n",
    "    \"Bat-CoV HKU8\": \"NC_010438\",\n",
    "    \"Bat-CoV HKU2\": \"NC_009988\",\n",
    "    \"Bat-CoV HKU5\": \"NC_009020\",\n",
    "    \"Bat-CoV RaTG13\": \"MN996532\",\n",
    "    \"Bat-CoV-ENT\": \"NC_003045\",\n",
    "    \n",
    "    # Other animals\n",
    "    \"Hedgehog-CoV 2012-174/GER/2012\": \"NC_039207\",\n",
    "    \"Pangolin-CoV MP789\": \"MT121216\",\n",
    "    \"Rabbit-CoV HKU14\": \"NC_017083\",\n",
    "    \"Duck-CoV isolate DK/GD/27/2014\": \"NC_048214\",\n",
    "    \"Feline infectious peritonitis virus\": \"NC_002306\",  # cat\n",
    "    \"Giraffe-CoV US/OH3/2003\": \"EF424623\",\n",
    "    \"Murine-CoV MHV/BHKR_lab/USA/icA59_L94P/2012\": \"KF268338\",  # mouse\n",
    "    \"Equine-CoV Obihiro12-2\": \"LC061274\",  # horse\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above is the list of coronaviruses and their corresponding NCBI accession codes that we will consider in this assignment. As in the first assignment, you can find the SARS-CoV-2 sequence in `data/SARS-CoV-2.fa`. This time, we will use the FASTA format to familiarize ourselves with the sequence file formats used in bioinformatics. You can easily read FASTA files using the `SeqIO.read` function from `biopython`.\n",
    "\n",
    "We will only use part of the sequence, and to assess similarity, consider the spike protein alone. In the first homework, we have already narrowed down our candidates for the transmembrane proteins, and we will assume that we know the location of the spike protein in SARS-CoV-2 (strand=1, start=21562, stop=25384). We need to find a part of the sequence that encodes for the spike protein in all coronaviruses listed above. We can inspect each sequence's features to locate the spike protein. For this assignment, we will iterate through gene coding regions (CDS) for each coronavirus and find the one that codes for the \"S\" gene. Some coronaviruses do not include such annotation but instead report on a \"spike protein\" in the `product` field."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sars_cov_2 = SeqIO.read(\"data/sars_cov_2.fa\", \"fasta\")\n",
    "sars_spike = sars_cov_2.seq[21562:25384].translate(to_stop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding all spike proteins\n",
    "spike_proteins = {}\n",
    "\n",
    "for code in accession_codes:\n",
    "    organism_id = accession_codes[code]\n",
    "    record = SeqIO.read(Entrez.efetch(db='nucleotide', id=organism_id, rettype='gbwithparts', retmode='text'), format='genbank')\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            if \"gene\" in feature.qualifiers:\n",
    "                if feature.qualifiers[\"gene\"] == ['S']:\n",
    "                    # print(feature.location.extract(record).seq) This gets the ORFs\n",
    "                    # print(feature.qualifiers[\"translation\"][0]) This gets the translated protein\n",
    "                    spike_proteins[code] = feature.qualifiers[\"translation\"][0]\n",
    "            elif \"product\" in feature.qualifiers:\n",
    "                if feature.qualifiers[\"product\"] == [\"spike protein\"]:\n",
    "                    spike_proteins[code] = feature.qualifiers[\"translation\"][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1: Global alignment\n",
    "\n",
    "**Task:** Use the template `global_alignment` function in the file `helper_functions.py` and implement the Needleman-Wunsch algorithm for global sequence alignment. Use a \"-\" character for indels. Score the alignments using the BLOSUM62 substitution matrix. You can find it in `biopython`.\n",
    "\n",
    "**[10 points]**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from helper_functions import global_alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2: Choosing our reference genomes\n",
    "\n",
    "**Task:** Identify the three closest relatives to SARS-CoV-2. Use the `global_alignment` function you have just implemented, inspect their alignment scores, and select the three most similar sequences to our spike protein sequence. These will serve as reference genomes in Problem 4 of this assignment and help us determine the function of ORFs.\n",
    "\n",
    "Save the full names of three coronaviruses with the closest matches to SARS-CoV-2's spike protein sequence as a Python list in the `three_closest_references` variable.\n",
    "\n",
    "**[6 points]**\n",
    "\n",
    "Then answer the following questions about global alignment and the three closest references.\n",
    "- When aligning spike protein sequences with the global alignment, we used the BLOSUM62 substitution matrix by default. Examine the similarity of the protein alignments (the frequency of matching amino acids) and discuss whether it would be better to use the BLOSUM90 or BLOSUM45 substitution matrix in our case.\n",
    "- The theory of evolution states that the SARS-CoV-2 virus must have evolved from some other virus in the past. Based on the three closest alignment references, is it more likely that SARS-CoV-2 jumped from another host organism or evolved from a human coronavirus? (We will explore this further in the following exercise.)\n",
    "\n",
    "Store your answers in the `blosum_selection` and `viral_origin` variables, respectively.\n",
    "\n",
    "**[4 points]**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load blosum62, blosum90 and blosum45 matrices\n",
    "from Bio.Align import substitution_matrices\n",
    "blosum62_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "blosum90_matrix = substitution_matrices.load(\"BLOSUM90\")\n",
    "blosum45_matrix = substitution_matrices.load(\"BLOSUM45\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Out of curiosity, calculate max score for sars_spike for each of those matrices\n",
    "_, _, max_blosum62 = global_alignment(sars_spike, sars_spike, lambda x, y: blosum62_matrix[x, y])\n",
    "_, _, max_blosum90 = global_alignment(sars_spike, sars_spike, lambda x, y: blosum90_matrix[x, y])\n",
    "_, _, max_blosum45 = global_alignment(sars_spike, sars_spike, lambda x, y: blosum45_matrix[x, y])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate other similarities, as well as save the alignments for calculations later\n",
    "similarities_62 = {}\n",
    "similarities_90 = {}\n",
    "similarities_45 = {}\n",
    "matching_62 = {}\n",
    "matching_90 = {}\n",
    "matching_45 = {}\n",
    "\n",
    "for protein in spike_proteins:\n",
    "    alignment_1_62, alignment_2_62, score62 = global_alignment(sars_spike, spike_proteins[protein], lambda x, y: blosum62_matrix[x, y])\n",
    "    alignment_1_90, alignment_2_90, score90 = global_alignment(sars_spike, spike_proteins[protein], lambda x, y: blosum90_matrix[x, y])\n",
    "    alignment_1_45, alignment_2_45, score45 = global_alignment(sars_spike, spike_proteins[protein], lambda x, y: blosum45_matrix[x, y])\n",
    "    similarities_62[protein] = score62\n",
    "    similarities_90[protein] = score90\n",
    "    similarities_45[protein] = score45\n",
    "    matching_62[protein] = (alignment_1_62, alignment_2_62)\n",
    "    matching_90[protein] = (alignment_1_90, alignment_2_90)\n",
    "    matching_45[protein] = (alignment_1_45, alignment_2_45)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|                   PROTEIN                   | BLOSUM62 | BLOSUM45 | BLOSUM90 |\n",
      "+---------------------------------------------+----------+----------+----------+\n",
      "|         (maximum possible score) SARS-CoV-2 |  6722.0  |  7960.0  |  7920.0  |\n",
      "----------------------------------------------+----------+----------+----------+\n",
      "|                                  Human-SARS |  5246.0  |  6262.0  |  5966.0  |\n",
      "|                                  Human-MERS |  1739.0  |  2230.0  |  1323.0  |\n",
      "|                             Human-HCoV-OC43 |  1617.0  |  2034.0  |  1139.0  |\n",
      "|                             Human-HCoV-229E |  1101.0  |  1394.0  |   541.0  |\n",
      "|                             Human-HCoV-NL63 |  1158.0  |  1477.0  |   534.0  |\n",
      "|                             Human-HCoV-HKU1 |  1544.0  |  1951.0  |  1017.0  |\n",
      "|                                Bat-CoV MOP1 |  1087.0  |  1381.0  |   432.0  |\n",
      "|                                Bat-CoV HKU8 |  1153.0  |  1505.0  |   572.0  |\n",
      "|                                Bat-CoV HKU2 |  1091.0  |  1414.0  |   513.0  |\n",
      "|                                Bat-CoV HKU5 |  1797.0  |  2255.0  |  1407.0  |\n",
      "|                              Bat-CoV RaTG13 |  6539.0  |  7741.0  |  7677.0  |\n",
      "|                                 Bat-CoV-ENT |  1629.0  |  2042.0  |  1136.0  |\n",
      "|              Hedgehog-CoV 2012-174/GER/2012 |  1694.0  |  2155.0  |  1272.0  |\n",
      "|                          Pangolin-CoV MP789 |  6085.0  |  7222.0  |  7081.0  |\n",
      "|                            Rabbit-CoV HKU14 |  1653.0  |  2072.0  |  1178.0  |\n",
      "|              Duck-CoV isolate DK/GD/27/2014 |  1124.0  |  1433.0  |   527.0  |\n",
      "|         Feline infectious peritonitis virus |  1150.0  |  1404.0  |   447.0  |\n",
      "|                     Giraffe-CoV US/OH3/2003 |  1623.0  |  2038.0  |  1134.0  |\n",
      "| Murine-CoV MHV/BHKR_lab/USA/icA59_L94P/2012 |  1719.0  |  2162.0  |  1280.0  |\n",
      "|                      Equine-CoV Obihiro12-2 |  1575.0  |  2012.0  |  1071.0  |\n"
     ]
    }
   ],
   "source": [
    "print('|                   PROTEIN                   | BLOSUM62 | BLOSUM45 | BLOSUM90 |')\n",
    "print('+---------------------------------------------+----------+----------+----------+')\n",
    "print(f'|         (maximum possible score) SARS-CoV-2 |  {max_blosum62}  |  {max_blosum45}  |  {max_blosum90}  |')\n",
    "print('----------------------------------------------+----------+----------+----------+')\n",
    "\n",
    "for protein in similarities_62:\n",
    "    a = (6 - len(str(similarities_90[protein]))) * ' '\n",
    "    b = (43 - len(protein)) * ' '\n",
    "\n",
    "    print(f'| {b}{protein} |  {similarities_62[protein]}  |  {similarities_45[protein]}  |  {a}{similarities_90[protein]}  |')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|                   PROTEIN                   |          BLOSUM62         |          BLOSUM45         |          BLOSUM90         |\n",
      "+---------------------------------------------+---------------------------+---------------------------+---------------------------+\n",
      "|                                             |    MATCH    |  NON-MATCH  |    MATCH    |  NON-MATCH  |    MATCH    |  NON-MATCH  |\n",
      "+---------------------------------------------+-------------+-------------+-------------+-------------+-------------+-------------+\n",
      "|                                  Human-SARS |  978 (0.76) |  302 (0.24) |  974 (0.76) |  303 (0.24) |  979 (0.77) |  299 (0.23) |\n",
      "|                                  Human-MERS |  487 (0.34) |  931 (0.66) |  476 (0.34) |  922 (0.66) |  477 (0.34) |  920 (0.66) |\n",
      "|                             Human-HCoV-OC43 |  469 (0.33) |  954 (0.67) |  454 (0.32) |  954 (0.68) |  456 (0.32) |  952 (0.68) |\n",
      "|                             Human-HCoV-229E |  409 (0.30) |  941 (0.70) |  406 (0.30) |  934 (0.70) |  398 (0.30) |  937 (0.70) |\n",
      "|                             Human-HCoV-NL63 |  430 (0.30) | 1016 (0.70) |  424 (0.29) | 1017 (0.71) |  427 (0.30) | 1012 (0.70) |\n",
      "|                             Human-HCoV-HKU1 |  458 (0.32) |  970 (0.68) |  446 (0.31) |  971 (0.69) |  455 (0.32) |  963 (0.68) |\n",
      "|                                Bat-CoV MOP1 |  431 (0.29) | 1033 (0.71) |  428 (0.29) | 1030 (0.71) |  430 (0.29) | 1028 (0.71) |\n",
      "|                                Bat-CoV HKU8 |  431 (0.30) | 1018 (0.70) |  411 (0.29) | 1016 (0.71) |  421 (0.29) | 1010 (0.71) |\n",
      "|                                Bat-CoV HKU2 |  385 (0.29) |  941 (0.71) |  379 (0.29) |  941 (0.71) |  376 (0.29) |  937 (0.71) |\n",
      "|                                Bat-CoV HKU5 |  500 (0.35) |  911 (0.65) |  482 (0.35) |  912 (0.65) |  501 (0.36) |  903 (0.64) |\n",
      "|                              Bat-CoV RaTG13 | 1240 (0.97) |   33 (0.03) | 1240 (0.97) |   33 (0.03) | 1240 (0.97) |   33 (0.03) |\n",
      "|                                 Bat-CoV-ENT |  471 (0.33) |  960 (0.67) |  461 (0.32) |  958 (0.68) |  464 (0.33) |  956 (0.67) |\n",
      "|              Hedgehog-CoV 2012-174/GER/2012 |  463 (0.33) |  936 (0.67) |  449 (0.32) |  935 (0.68) |  451 (0.33) |  929 (0.67) |\n",
      "|                          Pangolin-CoV MP789 | 1151 (0.90) |  126 (0.10) | 1150 (0.90) |  126 (0.10) | 1150 (0.90) |  126 (0.10) |\n",
      "|                            Rabbit-CoV HKU14 |  469 (0.33) |  953 (0.67) |  463 (0.33) |  952 (0.67) |  466 (0.33) |  950 (0.67) |\n",
      "|              Duck-CoV isolate DK/GD/27/2014 |  430 (0.31) |  947 (0.69) |  420 (0.31) |  940 (0.69) |  412 (0.30) |  947 (0.70) |\n",
      "|         Feline infectious peritonitis virus |  450 (0.30) | 1074 (0.70) |  442 (0.29) | 1071 (0.71) |  441 (0.29) | 1073 (0.71) |\n",
      "|                     Giraffe-CoV US/OH3/2003 |  466 (0.33) |  962 (0.67) |  455 (0.32) |  957 (0.68) |  454 (0.32) |  958 (0.68) |\n",
      "| Murine-CoV MHV/BHKR_lab/USA/icA59_L94P/2012 |  478 (0.34) |  930 (0.66) |  474 (0.34) |  926 (0.66) |  477 (0.34) |  925 (0.66) |\n",
      "|                      Equine-CoV Obihiro12-2 |  461 (0.32) |  963 (0.68) |  454 (0.32) |  961 (0.68) |  456 (0.32) |  959 (0.68) |\n"
     ]
    }
   ],
   "source": [
    "print('|                   PROTEIN                   |          BLOSUM62         |          BLOSUM45         |          BLOSUM90         |')\n",
    "print('+---------------------------------------------+---------------------------+---------------------------+---------------------------+')\n",
    "print('|                                             |    MATCH    |  NON-MATCH  |    MATCH    |  NON-MATCH  |    MATCH    |  NON-MATCH  |')\n",
    "print('+---------------------------------------------+-------------+-------------+-------------+-------------+-------------+-------------+')\n",
    "# Calculate how many matches, mismatches (no '-') is for each matrix\n",
    "for protein in matching_62:\n",
    "    match_62 = 0\n",
    "    match_45 = 0\n",
    "    match_90 = 0\n",
    "\n",
    "    al62_1, al62_2 = matching_62[protein]\n",
    "    n62 = len(al62_1)\n",
    "    for i in range(len(al62_1)):\n",
    "        match_62 += 1 if al62_1[i] == al62_2[i] else 0\n",
    "    mismatch_62 = n62 - match_62\n",
    "\n",
    "    al45_1, al45_2 = matching_45[protein]\n",
    "    n45 = len(al45_1)\n",
    "    for i in range(len(al45_1)):\n",
    "        match_45 += 1 if al45_1[i] == al45_2[i] else 0\n",
    "    mismatch_45 = n45 - match_45\n",
    "\n",
    "    al90_1, al90_2 = matching_90[protein]\n",
    "    n90 = len(al90_1)\n",
    "    for i in range(len(al90_1)):\n",
    "        match_90 += 1 if al90_1[i] == al90_2[i] else 0\n",
    "    mismatch_90 = n90 - match_90\n",
    "\n",
    "    # Addons to make print prettier\n",
    "    a62 = (4 - len(str(match_62))) * ' '\n",
    "    b62 = '' if len(str(round(match_62 / n62, 2))) == 4 else '0'\n",
    "    c62 = (4 - len(str(mismatch_62))) * ' '\n",
    "    d62 = '' if len(str(round(mismatch_62 / n62, 2))) == 4 else '0'\n",
    "    a45 = (4 - len(str(match_45))) * ' '\n",
    "    b45 = '' if len(str(round(match_45 / n45, 2))) == 4 else '0'\n",
    "    c45 = (4 - len(str(mismatch_45))) * ' '\n",
    "    d45 = '' if len(str(round(mismatch_45 / n45, 2))) == 4 else '0'\n",
    "    a90 = (4 - len(str(match_90))) * ' '\n",
    "    b90 = '' if len(str(round(match_90 / n90, 2))) == 4 else '0'\n",
    "    c90 = (4 - len(str(mismatch_90))) * ' '\n",
    "    d90 = '' if len(str(round(mismatch_90 / n90, 2))) == 4 else '0'\n",
    "    e = (43 - len(protein)) * ' '\n",
    "\n",
    "    print(f'| {e}{protein} | {a62}{match_62} ({round(match_62 / n62, 2)}{b62}) | {c62}{mismatch_62} ({round(mismatch_62 / n62, 2)}{d62}) | {a45}{match_45} ({round(match_45 / n45, 2)}{b45}) | {c45}{mismatch_45} ({round(mismatch_45 / n45, 2)}{d45}) | {a90}{match_90} ({round(match_90 / n90, 2)}{b90}) | {c90}{mismatch_90} ({round(mismatch_90 / n90, 2)}{d90}) |')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "three_closest_references = [  # Enter the three most closely-related viruses\n",
    "    'Bat-CoV RaTG13',\n",
    "    'Pangolin-CoV MP789',\n",
    "    'Human-SARS',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "blosum_selection = \"\"\"\n",
    "Would it be better to use BLOSUM90 or BLOSUM45 substitution matrix in our case.\n",
    "The frequency of matching amino acids is roughly the same for all three BLOSUM matrices (at most 0.01 difference). So we will compare the alignment scores we get from each of\n",
    "the three matrices. Comparing BLOSUM45 to BLOSUM62, we notice that the scores are higher across the board, so I would say that choosing between these two is just a matter of\n",
    "preference. But if we compare BLOSUM90 to BLOSUM62, we see that high scores get higher but low scores get lower (5246 -> 5966 for Human-SARS and 1694 -> 1272 for Hedgehog-CoV\n",
    "for example). So I would argue it would be beneficial to use BLOSUM90, since the gap between similar and different sequences is higher (Sort of like squaring makes small numbers\n",
    "smaller and large ones larger).\n",
    "NOTE: The match and non-match numbers in the table printed above are calculated comparing alignments, generated from global_alignment and not by comparing protein sequences.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_origin = \"\"\"\n",
    "Is it more likely that SARS-CoV-2 jumped from another host organism or evolved from a human coronavirus?\n",
    "Based on the results above, I would say that there is a high chance that SARS-CoV-2 has jumped from a Bat-Cov RaTG13. Looking at the numbers, the similarity scores are very high\n",
    "across all three BLOSUM matrices, with the frequency of matching amino acids being 97%.\n",
    "NOTE: Matching amino acids here are calculated comparing alignments, generated by global_alignment and not by comparing proteins.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MiniBLAST\n",
    "\n",
    "In the previous homework, your task was to find ORF candidates, which we then classified into transmembrane/non-transmembrane proteins using a naive approach based on hydrophobicity. A far more sophisticated approach for characterizing any genetic sequence is BLAST. BLAST stores a database of annotated genes (usually from various organisms) and then uses local alignment and heuristic search to find database genes with a similar sequence to a query one. Assuming that genes with similar protein sequences perform similar functions, we can infer the function of a query sequence by looking at the function of any matching genes returned by BLAST. It is also possible that BLAST will not find suitable matches, which may signal that a sequence might not be an actual gene or protein. In this problem, we will implement our simplified BLAST version and call it MiniBLAST. We will use MiniBLAST to determine whether each of our candidate ORFs is a gene. And if it is a gene, we will use the annotations to determine what the gene does.\n",
    "\n",
    "Please note that BLAST is a complicated, optimized, heuristic, state-of-the-art technique to obtain very good approximate solutions. BLAST can query thousands of sequences in seconds. Our implementation will be slightly less sophisticated and somewhat more constrained but conceptually similar to that of BLAST.\n",
    "\n",
    "## Problem 3: Local alignment\n",
    "\n",
    "**Task:** Use the template`local_alignment` function in `helper_functions.py` and implement the Smith-Waterman algorithm for local sequence alignment.\n",
    "\n",
    "**[10 points]**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from helper_functions import local_alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Finding homologous genes\n",
    "\n",
    "For this assignment, we have selected five promising SARS-CoV-2 ORFs we found in our first homework. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_candidates = {\n",
    "    \"ORF-1\": (1, 11995, 13483),\n",
    "    \"ORF-2\": (1, 26792, 27191),\n",
    "    \"ORF-3\": (1, 23650, 25384),\n",
    "    \"ORF-4\": (-1, 421, 667),\n",
    "    \"ORF-5\": (1, 9133, 13483),\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now use MiniBLAST to compare these ORFs to known, annotated genes from the three closely-related coronaviruses you have stored in `three_closest_references`. The goal is to find the best matching gene from the reference genomes and decide if the quality of the match is sufficient to claim that ORF is an actual gene. \n",
    "\n",
    "Hint: among the five ORFs included in the candidates, we have planted one that is not a true gene.\n",
    "\n",
    "**Task:** We start by building a reference database. For each of the three closely-related coronaviruses from Problem 2, extract the coding regions (CDS) from each genome, and convert them to protein sequences. Store the name of the virus each protein sequence came from and its function. In our case, the gene's function is evident from its name, e.g., \"spike protein\", \"ORF1ab\", and similar. Given `orf_candidates`, compute the local alignment to each protein sequence in our database. Inspect each alignment and its corresponding score, and decide whether the alignment is sufficiently good to assume they perform the same function.\n",
    "\n",
    "Save your answers into the `orf_matches` variable as indicated in the cell below. Each ORF should be assigned a *closest-organism*, reporting the reference virus with the closest match, and a *homologous-gene*, indicating which gene the ORF matched to. If two ORFs match the reference with the same score, report one of them. For ORFs with no match in the reference, use None for both fields.\n",
    "\n",
    "**[10 points]**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "closely_related = {\"Bat-CoV RaTG13\": \"MN996532\", \"Pangolin-CoV MP789\": \"MT121216\", \"Human-SARS\": \"NC_004718\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "mini_database = {}\n",
    "\n",
    "i = 0\n",
    "\n",
    "for code in closely_related:\n",
    "    mini_database[code] = []\n",
    "    organism_id = closely_related[code]\n",
    "    record = SeqIO.read(Entrez.efetch(db='nucleotide', id=organism_id, rettype='gbwithparts', retmode='text'), format='genbank')\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            mini_database[code].append((feature.qualifiers['product'][0], feature.qualifiers['translation'][0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "similarities = {}\n",
    "\n",
    "for orf in orf_candidates:\n",
    "    strand, start, stop = orf_candidates[orf]\n",
    "    protein = sars_cov_2.seq[start:stop].reverse_complement().translate(to_stop=True) if strand == -1 else sars_cov_2.seq[start:stop].translate(to_stop=True)\n",
    "    highest_score = 0\n",
    "    al_1, al_2, func, virus = '', '', '', ''\n",
    "\n",
    "    for code in mini_database:\n",
    "        for function, protein_db in mini_database[code]:\n",
    "            alignment_1, alignment_2, score = local_alignment(protein, protein_db, lambda x, y: blosum62_matrix[x, y])\n",
    "            if score > highest_score:\n",
    "                highest_score = score\n",
    "                al_1 = alignment_1\n",
    "                al_2 = alignment_2\n",
    "                func = function\n",
    "                virus = code\n",
    "\n",
    "    similarities[orf] = (highest_score, virus, func, al_1, al_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|  ORF  |  SCORE |    MATCH   | CLOSEST ORGANISM |    GENE FUNCTION   |\n",
      "+-------+--------+------------+------------------+--------------------+\n",
      "| ORF-1 | 2553.0 |  491 (1.0) |  Bat-CoV RaTG13  | orf1ab polyprotein |\n",
      "| ORF-2 |  678.0 |  132 (1.0) |  Bat-CoV RaTG13  |   membrane protein |\n",
      "| ORF-3 | 2999.0 |  575 (1.0) |  Bat-CoV RaTG13  | spike glycoprotein |\n",
      "| ORF-4 |   61.0 |   17 (0.2) |      Human-SARS  | ORF1ab polyprotein |\n",
      "| ORF-5 | 7569.0 | 1440 (1.0) |  Bat-CoV RaTG13  | orf1ab polyprotein |\n"
     ]
    }
   ],
   "source": [
    "print('|  ORF  |  SCORE |    MATCH   | CLOSEST ORGANISM |    GENE FUNCTION   |')\n",
    "print('+-------+--------+------------+------------------+--------------------+')\n",
    "\n",
    "for orf in similarities:\n",
    "    score, closest, gene, al_1, al_2 = similarities[orf]\n",
    "    match = 0\n",
    "    n = len(al_1)\n",
    "\n",
    "    for i in range(len(al_1)):\n",
    "        match += 1 if al_1[i] == al_2[i] else 0\n",
    "\n",
    "    a = (6 - len(str(score))) * ' '\n",
    "    b = (4 - len(str(match))) * ' '\n",
    "    c = (14 - len(closest)) * ' '\n",
    "    d = (18 - len(gene)) * ' '\n",
    "\n",
    "    print(f'| {orf} | {a}{score} | {b}{match} ({round(match / n, 2)}) |  {c}{closest}  | {d}{gene} |')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_matches = {\n",
    "    \"ORF-1\": {\n",
    "        # These are just example solutions. You have to replace them with the correct answers\n",
    "        \"closest-organism\": \"Bat-CoV RaTG13\", \n",
    "        \"homologous-gene\": \"orf1ab polyprotein\",\n",
    "    },\n",
    "    \"ORF-2\": {\n",
    "        \"closest-organism\": \"Bat-CoV RaTG13\",\n",
    "        \"homologous-gene\": \"membrane protein\",\n",
    "    },\n",
    "    \"ORF-3\": {\n",
    "        \"closest-organism\": \"Bat-CoV RaTG13\",\n",
    "        \"homologous-gene\": \"spike glycoprotein\",\n",
    "    },\n",
    "    \"ORF-4\": {\n",
    "        \"closest-organism\": None,\n",
    "        \"homologous-gene\": None,\n",
    "    },\n",
    "    \"ORF-5\": {\n",
    "        \"closest-organism\": \"Bat-CoV RaTG13\",\n",
    "        \"homologous-gene\": \"orf1ab polyprotein\",\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bonus Problem: Repurposing SARS-CoV drug treatments for SARS-CoV-2\n",
    "\n",
    "You might have noticed that the coronavirus we've been working with ends with the number 2. That means there must have been a SARS-CoV-1 at some point, right? Indeed there was. The SARS-CoV-1 virus caused a SARS outbreak in 2002-2004 and infected over 8,000 people from 29 countries, killing at least 774 people. These may seem like small numbers today, but at the time, the outbreak caused much panic because the virus had a much higher mortality rate than SARS-CoV-2. \n",
    "\n",
    "Several drugs were developed and tested for the treatment of SARS. Some proved effective. Since SARS-CoV-2 is closely related to SARS-CoV-1, we could repurpose some effective drugs for SARS-CoV-1. We will look at one drug in particular: EK1 binds to a specific region on the SARS-CoV-1 spike protein and prevents the virus from entering the cells, thus stopping the infection and, consequently, the spread of the virus. Assuming the spike protein in SARS-CoV-2 is similar enough to the spike protein in SARS-CoV-1, EK1 may be able to bind to the SARS-CoV-2 spike protein and subsequently stop the infection.\n",
    "\n",
    "EK1 recognizes a specific sequence (or motif) of aminoacid types: **_hpphcpc_** where **_h_** is a hydrophobic, **_p_** is a polar, and **_c_** is a charged amino acid. There are approximately six consecutive repeats of this motif in SARS-CoV-1, so we expect to find a similar pattern in the SARS-CoV-2 spike protein. The individual amino acid types in this motif are not equally important for the successful binding of EK1. For instance, it is much more important for the **_h_** amino acids to be in the correct spots than the **_p_** or **_c_** amino acids. It is also critical that none of the amino acids are missing in the spike protein, as this will prevent binding.\n",
    "\n",
    "Aminoacids and their corresponding types (**_h_**, **_p_**, or **_c_**) are provided in `data/aminoacid_properties.csv`.\n",
    "\n",
    "**Task:** Determine whether EK1 can bind to the SARS-CoV-2 spike protein. Based on the description of how EK1 binds to the spike protein in the paragraphs above, design a scoring function that will appropriately weigh amino acid matches/mismatches. Then, perform local alignment between the SARS-CoV-1 spike protein sequence and the EK1 target motif (**_hpphcpc_** repeated six times). Since EK1 was developed for SARS-CoV-1, we know there should be a valid binding motif somewhere along the SARS-CoV-1 spike protein. Use this fact to calibrate your scoring function. Can we find an EK1 binding motif in the SARS-CoV-2 spike protein? Plot the positions of the best alignments in both viruses. For each virus, plot a horizontal line representing the spike protein sequence and a bar indicating the location of the EK1 binding motif. Ensure the two genomes are aligned to see whether the binding motifs appear in the same general regions of the spike protein. Save your figure into `problem-bonus.png`.\n",
    "\n",
    "**[5 points]**\n",
    "\n",
    "Then answer the following questions:\n",
    "- Does the SARS-CoV-2 spike protein contain the EK1 binding motif?\n",
    "- Are SARS-CoV-1 and SARS-CoV-2 motifs comparable? What procentage of the aminoacids match between the motifs?\n",
    "- Since the introduction of EK1 as a potential COVID-19 treatment, the drug has seen further development. Does the treatment work on COVID-19 patients? Search through literature and provide at least two references.\n",
    "\n",
    "Store your answers in the `binding_motif_present`, `binding_motif_similarity` and `EK1_drug_efficacy` variables, respectively.\n",
    "\n",
    "**[5 points]**\n",
    "\n",
    "*Hint*: Our task is to design a scoring function that reflects the underlying process of binding EK1 to the spike protein. For instance, since no amino acids must be missing from the spike protein sequence, we can design our scoring function to disallow any indels in our alignment. We can achieve this by heavily penalizing indels. Similarly, if we know hydrophobic matches are more important than polar or charged matches, we can appropriately weigh matches and penalize mismatches between these types of proteins.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "binding_motif_present = \"\"\"\n",
    "Does the SARS-CoV-2 spike protein contain the EK1 binding motif?\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "binding_motif_similarity = \"\"\"\n",
    "Are SARS-CoV-1 and SARS-CoV-2 motifs comparable? What procentage of the aminoacids match between the motifs?\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "EK1_drug_efficacy = \"\"\"\n",
    "Does the treatment work on COVID-19 patients? Search through literature and provide at least two references.\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinf",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
